src/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs

```diff
@@ -4,21 +4,27 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Collections.ObjectModel;
+using System.Diagnostics;
 using System.Linq.Expressions;
 using System.Reflection;
 
 namespace System.Linq
 {
-    internal class EnumerableRewriter : OldExpressionVisitor
+    internal class EnumerableRewriter : ExpressionVisitor
```

Comment:
If you call `Execute` or `Execute<TResult>` on the linq-to-objects `IQueryProvider` implementation that will cause it to be created once per call. If you enumerate a linq-to-objects query then it is called once and the resulting `IEnumerable<T>` is cached.

The target dictionary is necessary for any rewriting of targets, as they must be replaced consistently (each reference of the same instance either not changed or always replaced with the same new instance). The type cache is memoising what can be relatively expensive. Both of them could perhaps be lazy-constructed though, and probably should be. In particular while the target dictionary is vital when touched on, it isn't touched on at all by the majority of cases (all currently working cases don't use it, as the old version doesn't support label expressions at all).